ls1intum/Artemis
# `Iris`: New message indicator

### Checklist
#### General


#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context


Show a red dot to indicate new messages. When user clicks on the chatbot icon, reposition user to the point that has new messages

### Description

- Whenever receiving a response from the server, increase `numNewMessages` by 1
- When close the chatbot dialog, reset `numNewMessages` to 0
- If chatbot is closed and `numNewMessages` > 0 then show the red dot
- If chatbot is opened, and `numNewMessage` > 0 then scroll to the unread message point. Otherwise, scroll to the bottom

### Steps for Testing

Prerequisites:
- 1 Student
- 1 Programming Exercise with Iris enabled

1. Log in to Artemis
2. Navigate to a Programming Exercise details page
3. Open chatbot
4. Throttle network and chat something, expect to see a "typing" indicator
5. Close the chatbot and wait until the response comes. Expect to see a red dot to indicate that there are unread messages
6. Open the chatbot and expect to be scrolled to the new message with a "unread message" line
7. Chat something and the "unread messages" line should stay at the same location
8. Close the chatbot and open it again and should be scrolled to the bottom, and "unread messages" line should be disappeared

Note: close the chatbot by either pressing the escape key, clicking on the "x" icon, or clicking on the chatbot icon. They should behave the same

#### Exam Mode Testing



### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage





| File                        | Lines           |Confirmation (assert/expect) |
|-----------------------------|-----------------|---|
| state-store.service.ts      | 95.34%		 |✅|
| exercise-chatbot.component.ts     | 90.56%		 |✅|
| exercise-chat-widget.component.ts        | 92.45%		 |✅|

### Screenshots



https://github.com/ls1intum/Artemis/assets/18320285/713c02ce-1a3e-4cd8-b789-256b877aed6a
